### PR TITLE
feat(annotations): update url on annotation selection

### PR DESCRIPTION
**Pull request by mickr**
Adding handler to update URL from annotation activity card, and call the callback from contentPreview to handle selecting annotation in Preview

- [x] Tests

### CODE REVIEW

**Comment by ConradJChan** [comment]
Few thoughts on this:

- What are your thoughts on instead of `handleAnnotationSelect` is that we could put a function on the AnnotatorContext similar to how we do connected components in redux:
```
export interface AnnotatorState {
    activeAnnotationId?: string | null;
    annotation?: object;
    action?: Action;
    error?: Error;
    setActiveAnnotationId: (id: string) => void;
}
```
In this way, we can avoid the `onAnnotationSelect` passing down through the component tree.

 - For the deep-linking to the annotation item, we need to handle it being initiated by either box-annotations or clicking the link in the annotation item -- what do you think about only relying on the box-annotations `annotations_active_change` event to trigger a `history.push|replace` to simplify the flow?

File: src/elements/common/annotator-context/withAnnotations.tsx [hunk 1]
```diff
@@ -21,8 +21,13 @@ export default function withAnnotations<P extends object>(
     class ComponentWithAnnotations extends React.Component<P & WithAnnotationsProps, AnnotatorState> {
         static displayName: string;
 
+        handleActiveChange = (annotationId: string | null): void => {
+            this.setState({ activeAnnotationId: annotationId });
+        };
+
         state: AnnotatorState = {
             activeAnnotationId: null,
+            setActiveAnnotationId: this.handleActiveChange,
```
Comment (ConradJChan) [inline src/elements/common/annotator-context/withAnnotations.tsx:30]: I think instead of directly setting it back to state, we may want to emit the `annotations_active_set` event back to the annotator, and let that naturally come back as a `annotations_active_change` event which will trigger `handleActiveChange`

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 2]
```diff
@@ -14,6 +14,7 @@ import AddTaskButton from './AddTaskButton';
 import API from '../../api';
 import messages from '../common/messages';
 import SidebarContent from './SidebarContent';
+import { AnnotatorState, withAnnotatorContext } from '../common/annotator-context';
```
Comment (ConradJChan) [inline src/elements/content-sidebar/ActivitySidebar.js:17]: Are these changes still needed here?

File: src/elements/content-sidebar/Sidebar.js [hunk 3]
```diff
@@ -102,13 +113,31 @@ class Sidebar extends React.Component<Props, State> {
...
     };
 
+    /**
+     * Handle annotation selection
+     *
+     * @param {string} annotationId - The id of the annotation being selected
+     * @return {void}
+     */
+
+    handleAnnotationSelect = (annotationId: string): void => {
```
Comment (ConradJChan) [inline src/elements/content-sidebar/Sidebar.js:134]: This may also be `null` so we should handle that case

File: src/elements/content-sidebar/ActivitySidebar.js (hunk 2 above)
Comment (mickr) [reply src/elements/content-sidebar/ActivitySidebar.js:17]: I could wrap the activity card with the context directly, but as of right now I'm passing it here to the activity feed.

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 4]
```diff
@@ -46,6 +47,7 @@ import './ActivitySidebar.scss';
 type ExternalProps = {
     activeFeedEntryId?: string,
     activeFeedEntryType?: FocusableFeedItemType,
+    annotatorState?: AnnotatorState,
```
Comment (ConradJChan) [inline src/elements/content-sidebar/ActivitySidebar.js:50]: Could instead do `& WithAnnotatorContextProps` at the end

File: src/elements/content-sidebar/Sidebar.js [hunk 5]
```diff
@@ -102,13 +113,35 @@ class Sidebar extends React.Component<Props, State> {
...
 
+    /**
+     * Handle annotation selection
+     *
+     * @param {string} annotationId - The id of the annotation being selected
+     * @return {void}
+     */
+
+    handleAnnotationSelect = (annotationId: string | null): void => {
+        if (!annotationId) {
```
Comment (ConradJChan) [inline src/elements/content-sidebar/Sidebar.js:135]: I think if we return here, we leave the url as a deep-linked annotation which wouldn't be accurate. Should it instead set the url to be `/`?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.scss [hunk 6]
```diff
@@ -72,6 +72,11 @@
             }
         }
 
+        .bcs-activity-feed-annotation-activity.bcs-is-focused {
+            padding-left: 21px;
```
Comment (ConradJChan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.scss:76]: Do we need to override the animation? I think this results in the focused annotation activity background-color fading back to transparent

File: src/elements/common/annotator-context/withAnnotations.tsx [hunk 7]
```diff
@@ -21,8 +21,13 @@ export default function withAnnotations<P extends object>(
     class ComponentWithAnnotations extends React.Component<P & WithAnnotationsProps, AnnotatorState> {
         static displayName: string;
 
+        handleActiveChange = (annotationId: string | null): void => {
```
Comment (jstoffan) [inline src/elements/common/annotator-context/withAnnotations.tsx:24]: Can this live next to the other handler, as before?

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 8]
```diff
@@ -594,6 +596,7 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
             activeFeedEntryId,
             activeFeedEntryType,
             onTaskView,
+            annotatorState = {},
```
Comment (jstoffan) [inline src/elements/content-sidebar/ActivitySidebar.js:599]: Default values should live in `defaultProps`.

File: src/elements/content-sidebar/Sidebar.js [hunk 9]
```diff
@@ -102,13 +113,35 @@ class Sidebar extends React.Component<Props, State> {
...
+            return;
+        }
+
+        const { history } = this.props;
+
+        const urlPrefix = this.getUrlPrefix(history.location.pathname);
+        history.replace(`/${urlPrefix}/annotations/${annotationId}`);
```
Comment (jstoffan) [inline src/elements/content-sidebar/Sidebar.js:142]: Do we want to `push` instead of `replace` here?

Comment (jstoffan) [reply src/elements/content-sidebar/Sidebar.js:142]: Is this URL going to suffice for our purposes? Will we be able to determine the version we need to load?

Comment (mickr) [reply src/elements/content-sidebar/Sidebar.js:142]: No, but I talked with the BE team and they have not worked through the versions url yet and what it will look like.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.scss (hunk 6 above)
Comment (mickr) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.scss:76]: I assumed that we would want to keep the focus behaviors similar to the other cards on the activity feed. I can follow up with design.

File: src/elements/common/annotator-context/withAnnotations.tsx (hunk 7 above)
Comment (mickr) [reply src/elements/common/annotator-context/withAnnotations.tsx:24]: I moved it here because TS was complaining that I attached it to the handler before it was declared. But some other feedback from Conrad for the event design will change this slightly.

File: src/elements/content-preview/ContentPreview.js [hunk 10]
```diff
@@ -83,6 +83,7 @@ type Props = {
     logoUrl?: string,
     measureRef: Function,
     messages?: StringMap,
+    onAnnotator: Function,
```
[the remark this answers is not in the document]
Comment (mickr) [reply src/elements/content-preview/ContentPreview.js:86]: good catch

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.scss (hunk 6 above)
Comment (mickr) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.scss:76]: Spoke with design and we should keep the background blue, can I follow up with another PR as this one is already getting pretty big?

File: src/elements/common/annotator-context/withAnnotations.tsx [hunk 11]
```diff
@@ -59,15 +75,20 @@ export default function withAnnotations<P extends object>(
             }
         };
 
+        handleOnAnnotator = (annotator: Annotator): void => {
+            this.setState({ annotator });
```
Comment (jstoffan) [inline src/elements/common/annotator-context/withAnnotations.tsx:79]: If we're passing the annotator instance around, it seems to defeat the purpose of the other portions of this component. Can we avoid storing the annotator instance?

File: src/elements/common/annotator-context/withAnnotations.tsx [hunk 12]
```diff
@@ -59,15 +75,23 @@ export default function withAnnotations<P extends object>(
             }
         };
 
+        handleOnAnnotator = (annotator: Annotator): void => {
+            this.annotator = annotator;
```
[the remark this answers is not in the document]
Comment (mickr) [reply src/elements/common/annotator-context/withAnnotations.tsx:79]: Yes, this PR is getting really big so we can follow up quickly

File: src/elements/common/annotator-context/__tests__/withAnnotatorContext.test.tsx [hunk 13]
```diff
@@ -38,5 +42,8 @@ describe('elements/common/annotator-context/withAnnotatorContext', () => {
             annotation: { foo: 'bar' },
             action: Action.CREATE_START,
         });
+        expect(
+            (wrappedComponent.props() as WrappedComponentProps<HTMLDivElement>).emitAnnotatorActiveChangeEvent,
```
Comment (jstoffan) [inline src/elements/common/annotator-context/__tests__/withAnnotatorContext.test.tsx:46]: Can we avoid this cast? Or at least avoid repeating it?

File: src/elements/common/annotator-context/withAnnotations.tsx [hunk 14]
```diff
@@ -1,37 +1,53 @@
...
         static displayName: string;
 
+        annotator: Annotator | undefined;
+
         state = defaultState;
 
+        emitActiveChangeEvent = (id: string | null) => {
+            const { annotator } = this;
+
+            if (!annotator || !annotator.emit) {
```
Comment (jstoffan) [inline src/elements/common/annotator-context/withAnnotations.tsx:44]: Is this an error condition? Should we handle it somehow other than swallowing it?

File: src/elements/common/annotator-context/withAnnotations.tsx [hunk 15]
```diff
@@ -59,15 +75,23 @@ export default function withAnnotations<P extends object>(
             }
         };
 
+        handleOnAnnotator = (annotator: Annotator): void => {
+            this.annotator = annotator;
+        };
+
         handlePreviewDestroy = (): void => {
             this.setState(defaultState);
+            this.annotator = undefined;
```
Comment (jstoffan) [inline src/elements/common/annotator-context/withAnnotations.tsx:84]: If this is a nullable property, use `null` rather than explicitly setting `undefined`.

File: src/elements/common/annotator-context/withAnnotations.tsx [hunk 16]
```diff
@@ -59,15 +75,23 @@ export default function withAnnotations<P extends object>(
...
         };
 
         render(): JSX.Element {
             return (
-                <AnnotatorContext.Provider value={this.state}>
+                <AnnotatorContext.Provider
+                    value={{ state: this.state, emitActiveChangeEvent: this.emitActiveChangeEvent }}
```
Comment (jstoffan) [inline src/elements/common/annotator-context/withAnnotations.tsx:90]: Alphabetize?

File: src/elements/common/annotator-context/withAnnotations.tsx [hunk 17]
```diff
@@ -59,15 +75,23 @@ export default function withAnnotations<P extends object>(
...
+                >
                     <WrappedComponent
                         {...this.props}
+                        onAnnotator={this.handleOnAnnotator}
```
Comment (jstoffan) [inline src/elements/common/annotator-context/withAnnotations.tsx:94]: `handleAnnotator` instead of `handleOnAnnotator`

File: src/elements/content-sidebar/__tests__/Sidebar.test.js [hunk 18]
```diff
@@ -12,7 +12,8 @@ jest.mock('../../common/async-load', () => () => 'LoadableComponent');
 jest.mock('../../../utils/LocalStore');
 
 describe('elements/content-sidebar/Sidebar', () => {
-    const getWrapper = props => shallow(<Sidebar file={{ id: 'id' }} location={{ pathname: '/' }} {...props} />);
+    const getWrapper = props =>
+        shallow(<Sidebar file={{ id: 'id' }} annotatorState={{}} location={{ pathname: '/' }} {...props} />);
```
Comment (jstoffan) [inline src/elements/content-sidebar/__tests__/Sidebar.test.js:16]: Should the Sidebar have a defaultProp value for annotatorState?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 19]
```diff
@@ -260,8 +261,8 @@ class ActivityFeed extends React.Component<Props, State> {
                             isDisabled={isDisabled}
                             currentUser={currentUser}
                             onTaskAssignmentUpdate={onTaskAssignmentUpdate}
-                            onAnnotationDelete={onAnnotationDelete}
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:263]: Is this move needed?

File: src/elements/common/annotator-context/withAnnotations.tsx (hunk 14 above)
Comment (mickr) [reply src/elements/common/annotator-context/withAnnotations.tsx:44]: No, the annotator is loaded through an event so it starts null or undefined, TS throws an error if you don't have a gate for it.

File: src/elements/common/annotator-context/types.ts [hunk 20]
```diff
@@ -6,13 +6,23 @@ export enum Action {
...
 export interface AnnotatorState {
     activeAnnotationId?: string | null;
     annotation?: object | null;
     action?: Action | null;
     error?: Error | null;
 }
 
+export interface AnnotatorContext {
+    state: AnnotatorState;
+    emitActiveChangeEvent: (id: string) => void;
```
[the remark this answers is not in the document]
Comment (mickr) [reply src/elements/common/annotator-context/types.ts:23]: I think it is more descriptive with this name, not sure if this detail needs to be abstracted away.

File: src/elements/content-sidebar/__tests__/Sidebar.test.js (hunk 18 above)
Comment (mickr) [reply src/elements/content-sidebar/__tests__/Sidebar.test.js:16]: Yes and I added it

File: src/elements/common/annotator-context/withAnnotations.tsx (hunk 17 above)
Comment (mickr) [reply src/elements/common/annotator-context/withAnnotations.tsx:94]: I can follow up on this

File: src/elements/common/annotator-context/types.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (ConradJChan) [reply src/elements/common/annotator-context/types.ts:23]: ok then at least make it `emitActiveSetEvent`?
